test/commentParserToESTree.js

```diff
@@ -767,9 +825,9 @@ description`
         descriptionEndLine: 1,
         descriptionLines: [
           {
-            delimiter: '/**',
+            delimiter: '*',
```

Comment:
If you need to get things to work with these changes, I am wondering if there is a bug in `estreeToString` because these particular changes seem odd. If `/**` is being added during stringification already due to the `delimiter` on the `JsdocBlock`, why does this first line have its own `*` delimiter too? What happens if this is instead set to the empty string instead for these (intro) delimiter lines?

Reply:
Cool. Re: the choice, I think preserving what was entered should be the priority. The AST is meant to be available for allowing manipulation that can bring linting-type changes.

Reply:
I agree that the intention behind this package is manipulation and not linting-type changes.

The changes required to exactly reproduce the input with no ESTree AST manipulation in the round trip processing is certainly expanding the scope of work for this PR. So my initial thought on implementing this is potentially adding two new fields to JSDocBlock and that is the opening / closing delimiter line separator. It will either be the empty string or `\n`. This will be inserted after `/**` before the first document line and before the closing `*/` after the last document line or tag.

This work will require more changes in various tests I'm sure, so is not going to be as "surgical" as the current PR. I'll also add round trip testing for a variety of comment blocks in `compact` spacing mode.

At this point I do want to bring up a change in test tool. I'm sure you are aware of `Vitest`. I also came from the C8 / Mocha angle before switching to Vitest. I temporarily used Vitest when making the changes I did in the current PR. It drastically speeds up testing and ease of evaluating temporary logging in testing through the Vitest UI available that perpetually runs tests quickly when source or tests change. I'm once again going to temporarily install it while I take a look at the work ahead. All of the existing tests work in Vitest. I don't want to rock the boat too much so to speak, but for a project like `jsdoccomment` Vitest is a great fit. I won't commit those changes and move everything back to C8 / Mocha when finished, but does provide great testing infrastructure for this kind of project. I just want to poll your interest in potentially adopting Vitest though. I think once you see it in action you'd be sold.

Reply:
> The changes required to exactly reproduce the input with no ESTree AST manipulation in the round trip processing is certainly expanding the scope of work for this PR. So my initial thought on implementing this is potentially adding two new fields to JsdocBlock and that is the opening / closing delimiter line separator. It will either be the empty string or `\n`. This will be inserted after `/**` before the first document line and before the closing `*/` after the last document line or tag.

What do you think of having `preserve` mode just include any newlines? I'm fine with it either way though.
 
> This work will require more changes in various tests I'm sure, so is not going to be as "surgical" as the current PR. I'll also add round trip testing for a variety of comment blocks in `compact` spacing mode.

Ok, sure.

> At this point I do want to bring up a change in test tool. I'm sure you are aware of `Vitest`. I also came from the C8 / Mocha angle before switching to Vitest. I temporarily used Vitest when making the changes I did in the current PR. It drastically speeds up testing and ease of evaluating temporary logging in testing through the Vitest UI available that perpetually runs tests quickly when source or tests change. I'm once again going to temporarily install it while I take a look at the work ahead. All of the existing tests work in Vitest. I don't want to rock the boat too much so to speak, but for a project like `jsdoccomment` Vitest is a great fit. I won't commit those changes and move everything back to C8 / Mocha when finished, but it does provide great testing infrastructure for this kind of project. I just want to poll your interest in potentially adopting Vitest though. I think once you see it in action you'd be sold.

Ok, can give it a try, sure.

Reply:
> What do you think of having `preserve` mode just include any newlines? I'm fine with it either way though.

If you can elaborate on the thought above I'd be glad to receive as much input before getting started. I am going to spend a bit of time to familiarize myself even more with the complete picture of the code base / engineering analysis. I'll likely get started this weekend.

The main goal that we discussed for the upcoming changes is exactly representing the input to output in the `parseComment` -> `commentParserToESTree` -> `estreeToString` round trip. This would be the case for the `preserve` mode. The new line / empty lines will be represented in the `documentLine` array where as the `compact` mode will not have empty `documentLine` entries. 

> > I just want to poll your interest in potentially adopting Vitest though. I think once you see it in action you'd be sold.
> 
> Ok, can give it a try, sure.

Just so there is no shock in test source structure. I normally split out data for data driven tests from the test source itself in a `/test/fixture` directory. To have Mocha or Vitest target the tests I use `.test.js` as the extension for the actual tests. I plan to create `/test/roundTrip.test.js` for the data driven round trip testing w/ the data defined in `/test/fixture/roundTrip.js`. The existing tests will be the same just with a `.test.js` extension.

One more procedural question I have relates to commit count. I try to keep PRs to a minimum, preferably one, commit. This expanded effort is a bit more work, so I'll be doing multiple commits. This can be alleviated with a [squash merge](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/about-merge-methods-on-github#squashing-your-merge-commits), so it might require a setting change or two.

Reply:
> > What do you think of having `preserve` mode just include any newlines? I'm fine with it either way though.
> 
> If you can elaborate on the thought above I'd be glad to receive as much input before getting started. I am going to spend a bit of time to familiarize myself even more with the complete picture of the code base / engineering analysis. I'll likely get started this weekend.

Sorry, I mean having the `description` just include the initial and final newlines when in `preserve` mode, rather than adding new fields for "opening / closing delimiter line separator".

> The main goal that we discussed for the upcoming changes is exactly representing the input to output in the `parseComment` -> `commentParserToESTree` -> `estreeToString` round trip. This would be the case for the `preserve` mode. The new line / empty lines will be represented in the `documentLine` array where as the `compact` mode will not have empty `documentLine` entries.

I guess you mean `descriptionLines` instead of `documentLine`? I see. I'm fine then with new fields for the opening/closing delimiter, so we don't need to add new entries to the beginning or end of `descriptionLines`.

> > > I just want to poll your interest in potentially adopting Vitest though. I think once you see it in action you'd be sold.
> > 
> > 
> > Ok, can give it a try, sure.
> 
> Just so there is no shock in test source structure. I normally split out data for data driven tests from the test source itself in a `/test/fixture` directory. To have Mocha or Vitest target the tests I use `.test.js` as the extension for the actual tests. I plan to create `/test/roundTrip.test.js` for the data driven round trip testing w/ the data defined in `/test/fixture/roundTrip.js`. The existing tests will be the same just with a `.test.js` extension.

Sure, that's fine.

> One more procedural question I have relates to commit count. I try to keep PRs to a minimum, preferably one, commit. This expanded effort is a bit more work, so I'll be doing multiple commits. This can be alleviated with a [squash merge](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/about-merge-methods-on-github#squashing-your-merge-commits), so it might require a setting change or two.

Sure, I can do a squash.

Reply:
OK.. @brettz9. Everything should be good to go now pending review.

The main core commits are [source changes](https://github.com/es-joy/jsdoccomment/pull/16/commits/9d59a950c9b69e5b74b37fbd96cd7d6affe7e2aa) and [test changes](https://github.com/es-joy/jsdoccomment/pull/16/commits/59b874d296008f4f3cdf767877c22eba43618d24)

I added extensive round trip testing as additional tests covering `parseComment` -> `commentParserToESTree` -> `estreeToString` for all relevant options (`compact` / `preserve` spacing and `preferRawType` / the default pratt type parser for estreeToString). Additionally, I test the `ast.description` and `tag.description` against the `descriptionLines` for each JsdocBlock and tags.

The main addition to the AST generated by `commentParserToESTree` is `delimiterLineBreak` and `preterminalLineBreak` which takes into account all possible variations of starting / closing delimiters. It is either `\n` or an `empty string`. 

I made an update to the README adding these properties.

I did revise `estreeToString` to directly traverse the AST instead of the initial visitor key `args` processing.

This is a non-trivial commit, but is well covered in new tests. Test coverage is now 100% on all files except the utilities in `jsdoccomment.js` which were not utilized in any of the changes.

My general recommendation is bumping to `0.5` as this essentially is a breaking change with the new properties in JsdocBlock and exact support input to output.

I'd be glad to help with any release notes as required.

Reply:
All right. I've reviewed everything and checked against `eslint-plugin-jsdoc`, so I think we're pretty much good to go.

Could you just add to `CHANGES.md`, crediting yourself as `@typhonrt` with the changes? I plan to then do a squash.

Reply:
Excellent, what version number; `0.43.0`?

Reply:
Yeah `0.43.0` sounds good...

Reply:
Got it updated. 

Yeah, once you get this merged I'll get another PR together for improved type declarations and API docs. Should have that ready tomorrow for review. One of the nice things you'll see is a more compact way to write `import types` in the JSDoc comments, bundled types, and accurate API docs.

Reply:
> One of the nice things you'll see is a more compact way to write `import types` in the JSDoc comments, bundled types, and accurate API docs.

This is still standard or?

Reply:
Yeah, it's all standard and you can review those changes separately. 

By default `esm-d-ts` uses the newer `Bundler` option for `moduleResolution` for TSC and you can reference `@typedefs` / type aliases via `import('.').ATYPEDEF` instead of `import('./index.js').ATYPEDEF`. It's small, but nice to reference things without an extension or explicit index file; definitely handy in larger codebases.

The bundled type declaration output is a big win for IDE tooling as all symbols are directly referenced and defined in a single `.d.ts` file rather than `import types` [peppered](https://github.com/es-joy/jsdoccomment/blob/main/dist/parseComment.d.ts#L34) in the raw output of `TSC` currently. Check the various current `.d.ts` files and see `import('./index.js').THING`. I use WebStorm and that won't parse / work correctly with intellisense (see below pict). This works correctly with bundled types.

WebStorm intellisense popover w/ current unbundled types (return type of `JsdocBlockWithInline` does not resolve):
![image](https://github.com/es-joy/jsdoccomment/assets/311473/26f91fcd-e770-4c8c-abf9-911343760fd9)